### PR TITLE
2Fa for admin panel

### PR DESCRIPTION
**Pull request by jarindr**
Issue/Task Number: #743 

# Overview

Add 2 FA for admin panel.

# Changes

- Add 2 FA for admin panel.
- Minor refactoring.

# Implementation Details

-

# Usage

Enable 2Fa in setting and test it out.

# Impact

Deploy as usual.

### CODE REVIEW

File: apps/frontend/assets/src/omg-2fa-modal/Enable2FaModal.js [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+          return (
+            <div key={backupCode} className='backup-item'>
+              {backupCode}
+            </div>
+          )
+        })}
+        <div>
+          <Button onClick={onOkBackupCode}>
+            {'Ok, I\'ve saved the backup codes'}
+            <br /> ( This will log you out )
```
Comment (nicholasmueller) [inline apps/frontend/assets/src/omg-2fa-modal/Enable2FaModal.js]: you can still close the modal at this step, and then you would navigate an empty app with 'invalid auth token' errors, so you should logout onrequestclose as well

File: apps/frontend/assets/src/omg-2fa-modal/Enable2FaModal.js [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+        <div>
+          Secret code: {_.get(secretCode, 'secret_2fa_code', 'loading code..')}
+        </div>
+        <QrCode data={secretCode && to2FaFormat(secretCode)} size={200} />
+        <Input
+          value={passcode}
+          onChange={e => setPasscode(e.target.value)}
+          normalPlaceholder='passcode...'
+        />
+        <Button loading={submitStatus === 'LOADING'}>
```
Comment (nicholasmueller) [inline apps/frontend/assets/src/omg-2fa-modal/Enable2FaModal.js]: if inputted code is wrong, button stays in loading state

File: apps/frontend/assets/src/omg-2fa-modal/Disable2FaModal.js [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+  }
+
+  const renderDisableSection = () => {
+    return (
+      <form onSubmit={onSubmit}>
+        <h4>Disable Your Two Factor Authentication</h4>
+        <Input
+          value={passcode}
+          onChange={e => setPasscode(e.target.value)}
+          normalPlaceholder='passcode...'
```
Comment (nicholasmueller) [inline apps/frontend/assets/src/omg-2fa-modal/Disable2FaModal.js]: maybe `2FA Token` is more clear? maybe people will think this is asking for their password

File: apps/frontend/assets/src/omg-2fa-modal/Disable2FaModal.js [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    } else {
+      setErrorText(result.error.description)
+      setSubmitStatus('FAILED')
+    }
+  }
+
+  const renderDisableSection = () => {
+    return (
+      <form onSubmit={onSubmit}>
+        <h4>Disable Your Two Factor Authentication</h4>
```
Comment (nicholasmueller) [inline apps/frontend/assets/src/omg-2fa-modal/Disable2FaModal.js]: grammar police 👮🏻‍♀️  "Disable Two Factor Authentication" is better i think

File: apps/frontend/assets/src/omg-2fa-modal/Enable2FaModal.js [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+  }
+  .backup-item {
+    padding: 5px;
+    display: inline-block;
+    width: 110px;
+  }
+`
+
+function CreateTwoFaModal ({ open, onRequestClose, history }) {
+  const dispatch = useDispatch()
```
Comment (nicholasmueller) [inline apps/frontend/assets/src/omg-2fa-modal/Enable2FaModal.js]: i love this, so much cleaner to pass this dispatch

File: apps/frontend/assets/src/omg-2fa-modal/Enable2FaModal.js [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+    )
+  }
+
+  const renderShowBackup = () => {
+    return (
+      <div className='backup-container'>
+        <h4>Please keep your backup code</h4>
+        <p>
+          You can print or download one-time use backup codes for times when
+          your phones are unavailable.
```
Comment (nicholasmueller) [inline apps/frontend/assets/src/omg-2fa-modal/Enable2FaModal.js]: grammar police 👮🏻‍♀️  "Please print or download these one-time use backup codes for times when your device is unavailable."

File: apps/frontend/assets/src/omg-alert/reducer.js [hunk 7]
```diff
@@ -32,76 +32,143 @@ export const alertsReducer = createReducer([], {
...
+    return [
+      ...state,
+      createAlertState('Created category successfully.', 'success')
+    ]
   },
   'ACCOUNT/ASSIGN_KEY/SUCCESS': state => {
-    return [...state, createAlertState('Assign key to account successfully.', 'success')]
+    return [
+      ...state,
+      createAlertState('Assign key to account successfully.', 'success')
```
Comment (nicholasmueller) [inline apps/frontend/assets/src/omg-alert/reducer.js]: grammar police 👮🏻‍♀️ => "Assigned key to..."

File: apps/frontend/assets/src/omg-2fa-modal/Enable2FaModal.js [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+
+  const onOkBackupCode = () => {
+    onRequestClose()
+    history.push('/login')
+  }
+
+  const renderCreateMode = () => {
+    return (
+      <form onSubmit={onSubmit}>
+        <h4>please scan the QR</h4>
```
Comment (nicholasmueller) [inline apps/frontend/assets/src/omg-2fa-modal/Enable2FaModal.js]: grammar police 👮🏻‍♀️ capitalize the header -> "Please..."

File: apps/frontend/assets/src/omg-2fa-modal/Enable2FaModal.js [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+          Enable Two Factor Authentication
+        </Button>
+      </form>
+    )
+  }
+
+  const renderShowBackup = () => {
+    return (
+      <div className='backup-container'>
+        <h4>Please keep your backup code</h4>
```
Comment (nicholasmueller) [inline apps/frontend/assets/src/omg-2fa-modal/Enable2FaModal.js]: grammar police 👮🏻‍♀️ "Please keep your backup codes"

File: apps/frontend/assets/src/omg-page-user-setting/SecuritySection.js [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+    margin: 20px 0;
+  }
+  .enable-two-fa {
+    margin-left: 20px;
+  }
+`
+function SecuritySeciton () {
+  const currentUser = useSelector(selectCurrentUser)
+  const dispatch = useDispatch()
+  const onClickEnable2Fa = useCallback(
```
Comment (nicholasmueller) [inline apps/frontend/assets/src/omg-page-user-setting/SecuritySection.js]: why do you need useCallback here?

Comment (jarindr) [reply apps/frontend/assets/src/omg-page-user-setting/SecuritySection.js]: not really necessary, just some perf for the page.

File: apps/frontend/assets/src/omg-2fa-modal/Disable2FaModal.js (hunk 3 above)
Comment (jarindr) [reply apps/frontend/assets/src/omg-2fa-modal/Disable2FaModal.js]: I'm using the same name as what backend use, can change the placeholder

**Pull request merged**